src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs

```diff
@@ -1887,6 +1891,29 @@ private static unsafe void AddToPollArray(Interop.PollEvent* arr, int arrLength,
                 bool success = false;
                 socket.InternalSafeHandle.DangerousAddRef(ref success);
                 int fd = (int)socket.InternalSafeHandle.DangerousGetHandle();
+
+                if (readCount > 0)
+                {
+                    // some platfoms like macOS do not like if there is duplication between real and error list.
+                    // To fix that we will search read list and if macthing descriptor exiost we will add events flags
+                    // instead of adding new entry to error list.
+                    int readIndex = 0;
+                    while (readIndex < readCount)
```

Comment:
Is this turning a linear operation into an N^2 operation?

Reply:
yes. This is where the assumption comes in that this is used from small number of sockets where it does not matter as much .... and the cost is only on platforms that are currently broken. (and use read and error list together)

Reply:
@GrabYourPitchforks, any concerns?

Reply:
I put up different implementation that avoids the N^2 problem. Please take another look @stephentoub